From 6dd258e2e1e3cf0661dfa9bebfb814507f845b55 Mon Sep 17 00:00:00 2001
From: Xin Wang <xiwang5@microsoft.com>
Date: Fri, 19 Sep 2025 14:40:46 +0800
Subject: [PATCH] Fix getattr AttributeError in multi-thread scenario

When multi-thread is used, the plugin loader may raise AttributeError
while getting "ActionModule" from an action plugin.

The reason is that cache is used while loading module. Before a module
is fully loaded, it is already put into the sys.modules cache. When another
thread tries to load the same module, it would load an incompletely loaded
module from the sys.modules cache. Then while getting "ActionModule"
attribute from this module, exception AttributeError could be raised.

Signed-off-by: Xin Wang <xiwang5@microsoft.com>
---
 lib/ansible/plugins/loader.py | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/lib/ansible/plugins/loader.py b/lib/ansible/plugins/loader.py
index b1f380558e..af9ffa2a23 100644
--- a/lib/ansible/plugins/loader.py
+++ b/lib/ansible/plugins/loader.py
@@ -934,14 +934,14 @@ class PluginLoader:
             spec = importlib.util.spec_from_file_location(to_native(python_module_name), to_native(path))
             module = importlib.util.module_from_spec(spec)

-            # mimic import machinery; make the module-being-loaded available in sys.modules during import
-            # and remove if there's a failure...
-            sys.modules[python_module_name] = module
-
             try:
                 spec.loader.exec_module(module)
+                # mimic import machinery; make the module-being-loaded available in sys.modules during import
+                # and remove if there's a failure...
+                sys.modules[python_module_name] = module
             except Exception:
-                del sys.modules[python_module_name]
+                if python_module_name in sys.modules:
+                    del sys.modules[python_module_name]
                 raise

         return module
--
2.51.0.windows.1
